File: packages/firebase_app_check/firebase_app_check/ios/firebase_app_check/Sources/firebase_app_check/FLTTokenRefreshStreamHandler.m
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FLTTokenRefreshStreamHandler.h"
#import "FLTFirebaseAppCheckPlugin.h"

const NSNotificationName kNotififactionEvent = @"FIRAppCheckAppCheckTokenDidChangeNotification";

NSString *const kTokenKey = @"FIRAppCheckTokenNotificationKey";

@implementation FLTTokenRefreshStreamHandler {
  id _observer;
}

- (FlutterError *)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)events {
  _observer =
      [NSNotificationCenter.defaultCenter addObserverForName:kNotififactionEvent
                                                      object:nil
                                                       queue:nil
                                                  usingBlock:^(NSNotification *_Nonnull note) {
                                                    NSString *token = note.userInfo[kTokenKey];

                                                    events(@{@"token" : token});
                                                  }];

  return nil;
}

- (FlutterError *)onCancelWithArguments:(id)arguments {
  [NSNotificationCenter.defaultCenter removeObserver:_observer];
  return nil;
}

@end

File: packages/firebase_app_check/firebase_app_check/ios/firebase_app_check/Sources/firebase_app_check/FLTTokenRefreshStreamHandler.swift
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

#if TARGET_OS_OSX
import FlutterMacOS
#else
import Flutter
#endif

private let kNotificationEvent = NSNotification.Name("FIRAppCheckAppCheckTokenDidChangeNotification")
private let kTokenKey = "FIRAppCheckTokenNotificationKey"

class FLTTokenRefreshStreamHandler: NSObject, FlutterStreamHandler {
    private var observer: NSObjectProtocol?
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        observer = NotificationCenter.default.addObserver(
            forName: kNotificationEvent,
            object: nil,
            queue: nil
        ) { notification in
            if let token = notification.userInfo?[kTokenKey] as? String {
                events(["token": token])
            }
        }
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        if let observer = observer {
            NotificationCenter.default.removeObserver(observer)
            self.observer = nil
        }
        return nil
    }
} 